docs: executing jupyter notebooks with py3.10 fails

Opening this for reference. For the purpose of building docs we don't need a higher python version than current one (3.8).
```bash
Built target doxygen
Converting jupyter notebooks to html
  NOTE: --execute requested
/home/runner/work/nrn/nrn/docs/tutorials
[NbConvertApp] Converting notebook ball-and-stick-1.ipynb to html
Warning: no DISPLAY environment variable.
--No graphics will be displayed.
[NbConvertApp] ERROR | Kernel died while waiting for execute reply.
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/client.py", line 618, in _async_poll_for_reply
    msg = await ensure_async(self.kc.shell_channel.get_msg(timeout=new_timeout))
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/util.py", line 96, in ensure_async
    result = await obj
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/jupyter_client/channels.py", line 224, in get_msg
    ready = await self.socket.poll(timeout)
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/client.py", line 830, in async_execute_cell
    exec_reply = await self.task_poll_for_reply
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.2/x64/bin/jupyter-nbconvert", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/jupyter_core/application.py", line 264, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/nbconvertapp.py", line 361, in start
    self.convert_notebooks()
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/nbconvertapp.py", line 533, in convert_notebooks
    self.convert_single_notebook(notebook_filename)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/nbconvertapp.py", line 498, in convert_single_notebook
    output, resources = self.export_single_notebook(notebook_filename, resources, input_buffer=input_buffer)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/nbconvertapp.py", line 427, in export_single_notebook
    output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/exporter.py", line 190, in from_filename
    return self.from_file(f, resources=resources, **kw)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/exporter.py", line 208, in from_file
    return self.from_notebook_node(nbformat.read(file_stream, as_version=4), resources=resources, **kw)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/html.py", line 145, in from_notebook_node
    return super().from_notebook_node(nb, resources, **kw)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/templateexporter.py", line 376, in from_notebook_node
    nb_copy, resources = super().from_notebook_node(nb, resources, **kw)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/exporter.py", line 147, in from_notebook_node
    nb_copy, resources = self._preprocess(nb_copy, resources)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/exporters/exporter.py", line 334, in _preprocess
    nbc, resc = preprocessor(nbc, resc)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/preprocessors/base.py", line 47, in __call__
    return self.preprocess(nb, resources)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/preprocessors/execute.py", line 84, in preprocess
    self.preprocess_cell(cell, resources, index)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbconvert/preprocessors/execute.py", line 105, in preprocess_cell
    cell = self.execute_cell(cell, index, store_history=True)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/util.py", line 84, in wrapped
    return just_run(coro(*args, **kwargs))
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/util.py", line 62, in just_run
    return loop.run_until_complete(coro)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/nbclient/client.py", line 834, in async_execute_cell
    raise DeadKernelError("Kernel died")
nbclient.exceptions.DeadKernelError: Kernel died
make[7]: *** [CMakeFiles/notebooks] Error 1
make[6]: *** [CMakeFiles/notebooks.dir/all] Error 2
make[5]: *** [CMakeFiles/notebooks.dir/rule] Error 2
CMakeFiles/notebooks.dir/build.make:70: recipe for target 'CMakeFiles/notebooks' failed
CMakeFiles/Makefile2:238: recipe for target 'CMakeFiles/notebooks.dir/all' failed
CMakeFiles/Makefile2:245: recipe for target 'CMakeFiles/notebooks.dir/rule' failed
make[4]: *** [notebooks] Error 2
Makefile:182: recipe for target 'notebooks' failed
make[3]: *** [CMakeFiles/docs] Error 2
CMakeFiles/docs.dir/build.make:70: recipe for target 'CMakeFiles/docs' failed
make[2]: *** [CMakeFiles/docs.dir/all] Error 2
CMakeFiles/Makefile2:316: recipe for target 'CMakeFiles/docs.dir/all' failed
make[1]: *** [CMakeFiles/docs.dir/rule] Error 2
CMakeFiles/Makefile2:323: recipe for target 'CMakeFiles/docs.dir/rule' failed
make: *** [docs] Error 2
Makefile:221: recipe for target 'docs' failed
Status : FAIL. Log:
None
Traceback (most recent call last):
  File "/home/runner/work/nrn/nrn/setup.py", line 481, in <module>
    setup_package()
  File "/home/runner/work/nrn/nrn/setup.py", line 424, in setup_package
    setup(
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/runner/work/nrn/nrn/setup.py", line 299, in run
    self.run_command("build_ext")
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/runner/work/nrn/nrn/setup.py", line 147, in run
    self._run_cmake(ext)  # Build cmake project
  File "/home/runner/work/nrn/nrn/setup.py", line 234, in _run_cmake
    subprocess.check_call(
  File "/opt/hostedtoolcache/Python/3.10.2/x64/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', 'docs']' returned non-zero exit status 2.
Error: Process completed with exit code 1.

```

--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -15,9 +15,7 @@ on:
       - release/**
 
 env:
-  # Python 3.10 fails executing jupyter notebooks. Don't need 3.10 for building docs anyway.
-  # see https://github.com/neuronsimulator/nrn/issues/1612
-  DEFAULT_PY_VERSION: '3.8'
+  DEFAULT_PY_VERSION: '3.10'
 
 jobs:
   documentation:

--- a/share/lib/python/neuron/__init__.py
+++ b/share/lib/python/neuron/__init__.py
@@ -632,15 +632,15 @@ def nrn_dll(printpath=False):
         be used with care.
     """
     import ctypes
-    import os
-    import platform
     import glob
+    import os
+    import sys
 
     try:
         # extended? if there is a __file__, then use that
         if printpath:
             print("hoc.__file__ %s" % _original_hoc_file)
-        the_dll = ctypes.cdll[_original_hoc_file]
+        the_dll = ctypes.pydll[_original_hoc_file]
         return the_dll
     except:
         pass
@@ -657,7 +657,7 @@ def nrn_dll(printpath=False):
         dlls = glob.glob(base_path + "*.*")
         for dll in dlls:
             try:
-                the_dll = ctypes.cdll[dll]
+                the_dll = ctypes.pydll[dll]
                 if printpath:
                     print(dll)
                 return the_dll
@@ -672,7 +672,7 @@ def nrn_dll(printpath=False):
         dlls = glob.glob(base_path + "*" + extension)
         for dll in dlls:
             try:
-                the_dll = ctypes.cdll[dll]
+                the_dll = ctypes.pydll[dll]
                 if printpath:
                     print(dll)
                 success = True

--- a/src/neuronmusic/nrnmusic.h
+++ b/src/neuronmusic/nrnmusic.h
@@ -3,7 +3,7 @@
 #if defined(NO_PYTHON_H)
 typedef struct _object PyObject;
 #else
-#include <Python.h>
+#include <nrnwrap_Python.h>
 #endif
 #include <music.hh>
 

--- a/src/nrnpython/nrnpython.h
+++ b/src/nrnpython/nrnpython.h
@@ -17,10 +17,6 @@
 #if defined(USE_PYTHON)
 #undef _POSIX_C_SOURCE
 #undef _XOPEN_SOURCE
-#if defined(__MINGW32__)
-//at least a problem with g++6.3.0
-#define _hypot hypot
-#endif
 #include <nrnwrap_Python.h>
 
 #endif /*USE_PYTHON*/

--- a/src/nrnpython/nrnwrap_Python.h
+++ b/src/nrnpython/nrnwrap_Python.h
@@ -1,7 +1,4 @@
-#ifndef nrnwrap_Python_h
-#define nrnwrap_Python_h
-/* avoid "redefined" warnings due to inconsistent configure HAVE_xxx */
-
+#pragma once
 #undef HAVE_PUTENV
 #undef HAVE_FTIME
 #undef HAVE_PROTOTYPES
@@ -9,6 +6,9 @@
 #undef _hypot
 #define _hypot hypot
 #endif
+// https://docs.python.org/3/c-api/intro.html#include-files states: It is
+// recommended to always define PY_SSIZE_T_CLEAN before including Python.h.
+#ifndef PY_SSIZE_T_CLEAN
+#define PY_SSIZE_T_CLEAN
+#endif
 #include <Python.h>
-
-#endif /* nrnwrap_Python_h */

